### ISSUE TITLE
Map card: style is unstable

### ISSUE THREAD

**Issue by ildar170975** [thomasloven/lovelace-card-mod#224]
My Home Assistant version: 2022.9.4

My lovelace configuration method (GUI or yaml): storage/yaml

What I am doing:
Styling a Map card.
I assigned an `entity_picture` (png with transparent background) to some tracker, now I want to remove a white circle around the picture & background color.
([initially it was a task about displaying ISS](https://community.home-assistant.io/t/card-mod-add-css-styles-to-any-lovelace-card/120744/3649?u=ildar_gabdullin)).

Here is a detailed path to the styled elements:
```
type: map
entities:
  - device_tracker.samsung_p6800_composite
  - device_tracker.ipad_air_2_composite
card_mod:
  style:
    ha-map:
      $:
        div#map:
          .leaflet-pane.leaflet-map-pane:
            .leaflet-pane.leaflet-marker-pane:
              .leaflet-marker-icon:
                ha-entity-marker:
                  $: |
                    .marker {
                      border: none !important;
                      background-color: transparent !important;
                    }
```
What I expect to see - the marker is with a transparent background & w/o a border:
![image](https://user-images.githubusercontent.com/71872483/190876287-6e6aa181-fe36-4121-bec5-21b15a8a9acb.png)

This works - until I refresh a page; then I see the not styled picture:
![image](https://user-images.githubusercontent.com/71872483/190876252-94931e8a-77bc-4a91-b1dd-92eb3924eee4.png)

Seems that elements are updated periodically and loosing their style.

Error messages from the browser console:
None
---

**By putting an X in the boxes ([]) below, I indicate that I:**

- [x] Understand that this is a channel for reporting bugs, not a support forum (https://community.home-assistant.io/).

- [x] Have made sure I am using the latest version of the plugin.

- [x] Have followed the troubleshooting steps of the "Common Problems" section of https://github.com/thomasloven/hass-config/wiki/Lovelace-Plugins.

- [x] Understand that leaving one or more boxes unticked or failure to follow the template above may increase the time required to handle my bug-report, or cause it to be closed without further action.

**Comment by kulmegil** [comment]
I'm surprised this has worked in the first place. 

Map elements (markers and other shapes) are controlled by [LeafletJS](https://leafletjs.com/) library lives outside of LitElement lifecycle - that is the component used mostly in Lovelace. The elements are rendered (and can possibly load too) dynamically as user interacts (moves, zooms, changes layers) of the map. Styling them using card-mod selectors is way above it's pay grade.

---
The best way would be to append styles to ALL `<ha-entity-marker>` elements.
Here is my lame attempt at wrinding an override:
```js
customElements.whenDefined('ha-entity-marker').then(() => {
  const EntityMarker = customElements.get('ha-entity-marker');
  const { html, css } = EntityMarker.prototype;
  
  // defined added style
  const newStyle = css`
  .marker {
    border: none !important;
    background-color: transparent !important;}`;
  
  const newStyles = [].concat(EntityMarker.styles, newStyle);
  // The LitElement class has already been finalized, but hopefully not yet instantiated
  // guess I'll just do it the hard way
  Object.defineProperty(EntityMarker, 'styles',        {value: newStyles, configurable: true, enumerable: false});
  Object.defineProperty(EntityMarker, 'elementStyles', {value: newStyles, configurable: true, enumerable: false});
});
```
Add it as [lovelace resource](https://www.home-assistant.io/lovelace/dashboards/#resources)
This will affect all markers on all map cards.  
It will also affect `/map` sidebar (depending if any dashboard was loaded after last page refresh)

**Comment by ildar170975** [comment]
@kulmegil 
Unfortunately, it is not stable.
1. Added the js code as a resource.
2. Open Map page - styles are applied:
![image](https://user-images.githubusercontent.com/71872483/193158482-d6bf8aca-5882-4fd9-aa8b-d0d930106e05.png)
3. Refresh (F5).
4. Styles are NOT applied:
![Uploading image.png…]()

**Comment by kulmegil** [comment]
@ildar170975
it's expected. I thought you don't use sidebar map since it's not very customizable and card_mod can be used only inside lovelace dashboards. Anyway in that case load is as [frontend module](https://www.home-assistant.io/integrations/frontend/#extra_module_url) (like card_mod) and it should _be stable_ everywhere

note: 
The general difference is that frontend modules are always loaded with HA interface, while resources when dashboard is selected. Also in the past I had experience problems with frontend modules not updating nicely in mobile_app. If You ever encounter this try changing module path or filename.

**Comment by ildar170975** [comment]
2024.10
For the record - the "card-mod way" works stable:
```
  - type: map
    entities:
      - device_tracker.ipad_air_2_accum
    card_mod:
      style:
        ha-map $ ha-entity-marker $: |
          .marker {
            border: 1px solid red !important;
            background-color: yellow !important;
          }
```
![image](https://github.com/user-attachments/assets/c29b3348-b7a0-4a97-acba-fb9da4a68b4c)

The style is persistent after F5.
Tested in Chrome+Win, in iOS 15.x (iPad).

**Comment by kulmegil** [comment]
@ildar170975

It appears so. 

Additionally it's possible to style individual entities via selectors now:
```yaml
type: map
entities:
  - entity: person.trump
  - entity: person.kamala
  - entity: person.joe
card_mod:
  style:
    ha-map$:
      ha-entity-marker[entity-id="person.trump"]$: |
        .marker { --ha-marker-shadow-color: var(--success-color); }
      ha-entity-marker[entity-id="person.kamala"]$: |
        .marker { --ha-marker-shadow-color: var(--error-color); }
      ha-entity-marker[entity-id="person.joe"]$: |
        .marker { --ha-marker-shadow-color: var(--amber-color); }
    ha-map$ ha-entity-marker$: |
      .marker {
        border-color: var(--ha-marker-shadow-color) !important;
        box-shadow: 
          -2px -2px 5px var(--ha-marker-shadow-color, black),
           2px -2px 5px var(--ha-marker-shadow-color, black),
           2px  2px 5px var(--ha-marker-shadow-color, black),
          -2px  2px 5px var(--ha-marker-shadow-color, black);
      }
```
![map](https://github.com/user-attachments/assets/ce1b03ff-f639-46e9-ba3b-eac71fa0ec94)

ps I have simplified code in my example above since it was "overengineered".
but it's recommended to use normal card_mod config with map now.

**Comment by ildar170975** [comment]
@kulmegil 
Thanks for coming back with precious remarks, appreciated!

As for the JS code you posted above (before your latest "simplification"):
I successfully used this approach for quick testing my PRs for HA frontend & other useful/useless tricks.
But is some cases I failed:
1. Either a `const { html, css } = UIelement.prototype;` gave an error.
2. Or `const newStyles = [].concat(SomeElement.styles, newStyle);` caused concatenating "empty" `styles` & `newStyles` - like that `SomeElement` dos not contain any `styles`.

I understand that this is an off-topic here; so, what could cause these issues?
